### PR TITLE
fix: Association mixin bug with unique false 

### PR DESCRIPTION
**Pull request by Necromansr**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Fixed association mixin bug `add` which ignores `unique: false` constraint. 
This pull request relates and fixes existing issue #14840

Previously `add` mixin completely ignores `unique: false` . 

In case where we want to implement own unique constraint/extend existing one
 we should use [`unique: false`](https://sequelize.org/docs/v6/core-concepts/assocs/#options-2) inside `belongsToMany` association. 
If we do this we expect that we can establish relation between 2 entities more than once.
Previously we could reach this only inside parallel querying:
```javascript
await Promise.all([
 a.addB(a, { through: { flag: false } }),
 a.addB(a, { through: { flag: true } })
]);
console.log((await a.getBs()).length); // will display 2
```
Which obviously gives us expected result only due to race condition (which is not good).

And If we do the same in sequential manner
```javascript
await a.addB(a, { through: { flag: false } });
await a.addB(a, { through: { flag: true } })


console.log((await a.getBs()).length); // will display 1
```
we will get different unexpected result.
The solution is to extend exising `add` method inside `belongs-to-many.ts` 
with additional if-statement check of `unique: false` `this.through` object.

### CODE REVIEW

**Comment by sdepold** [comment]
Thanks for your change :) Can you add a test for this as well?

**Comment by JaoodxD** [comment]
> Thanks for your change :) Can you add a test for this as well?

Added chai tests for this case.

File: src/associations/belongs-to-many.ts [hunk 1]
```diff
@@ -646,17 +646,21 @@ Add your own primary key to the through model, on different attributes than the
...
-        [this.otherKey]: newInstances.map(newInstance => newInstance.get(this.targetKey)),
-        ...this.through.scope,
-      },
-      // force this option to be false, in case the user enabled
-      rejectOnEmpty: false,
-    });
+    let currentRows: any[] = [];
+    if (this.through?.unique ?? true) {
+      currentRows = await this.through.model.findAll({
+        // const currentRows = await this.through.model.findAll({
```
Comment (WikiRik) [inline src/associations/belongs-to-many.ts:652]:
```suggestion
```

File: test/integration/associations/belongs-to-many.test.js [hunk 2]
```diff
@@ -3538,4 +3538,97 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
...
+      });
+    });
+
+    it('should create new relations in parallel', async function () {
+      await this.sequelize.sync({ force: true });
+
+      const { Order, Good } = this;
+
+      const order = await Order.create();
+      const good = await Good.create({ name: 'Coca-Cola' });
```
Comment (WikiRik) [inline test/integration/associations/belongs-to-many.test.js:3580]:
```suggestion
      const good = await Good.create({ name: 'Drink' });
```
Other suggestions are fine as well but I'd prefer not to include brand names

File: test/integration/associations/belongs-to-many.test.js [hunk 3]
```diff
@@ -3538,4 +3538,97 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
...
+        .equal(secondGood.orderGood.isUpsell);
+    });
+
+    it('should create new relations sequentialy', async function () {
+      await this.sequelize.sync({ force: true });
+
+      const { Order, Good } = this;
+
+      const order = await Order.create();
+      const good = await Good.create({ name: 'Coca-Cola' });
```
Comment (WikiRik) [inline test/integration/associations/belongs-to-many.test.js:3611]:
```suggestion
      const good = await Good.create({ name: 'Drink' });
```
Same as above

**Comment by JaoodxD** [comment]
@WikiRik your fixes make sense, ty.
Waiting for the review to become a contributor of this community 😉.

**Pull request merged**